src/lyrics/lyrics-qt.cc

```diff
@@ -0,0 +1,304 @@
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+ * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <QApplication>
+#include <QContextMenuEvent>
+#include <QDesktopServices>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QMenu>
+#include <QRegularExpression>
+#include <QTextEdit>
+
+#include <libaudcore/drct.h>
+#include <libaudcore/hook.h>
+#include <libaudcore/plugin.h>
+#include <libaudcore/plugins.h>
+
+#include "lyrics.h"
+#include "preferences.h"
+
+class TextEdit : public QTextEdit
+{
+public:
+    TextEdit (QWidget * parent = nullptr) : QTextEdit (parent) {}
+
+protected:
+    void contextMenuEvent (QContextMenuEvent * event);
+};
+
+class LyricsQt : public GeneralPlugin
+{
+public:
+    static const char * const defaults[];
+    static const PluginPreferences prefs;
+    static const PreferencesWidget widgets[];
+    static constexpr PluginInfo info = {
+        N_("Lyrics"),
+        PACKAGE,
+        nullptr, // about
+        & prefs,
+        PluginQtOnly
+    };
+
+    constexpr LyricsQt () : GeneralPlugin (info, false) {}
+
+    bool init ();
+    void * get_qt_widget ();
+};
+
+EXPORT LyricsQt aud_plugin_instance;
+
+const char * const LyricsQt::defaults[] = {
+    "remote-source", "lyrics.ovh",
+    "enable-file-provider", "TRUE",
+    "enable-cache", "TRUE",
+    "split-title-on-chars", "FALSE",
+    "split-on-chars", "-",
+    "truncate-fields-on-chars", "FALSE",
+    "truncate-on-chars", "|",
+    "use-embedded", "TRUE",
+    nullptr
+};
+
+const PreferencesWidget LyricsQt::widgets[] = {
+    WidgetLabel (N_("<b>General</b>")),
+    WidgetCheck (N_("Split title into artist and title on chars"),
+        WidgetBool (CFG_SECTION, "split-title-on-chars")),
+    WidgetTable ({{split_elements}}, WIDGET_CHILD),
+    WidgetLabel (N_("<b>Sources</b>")),
+    WidgetCheck (N_("Use embedded lyrics (from Lyrics tag)"),
+        WidgetBool (CFG_SECTION, "use-embedded")),
+    WidgetCombo (N_("Fetch lyrics from internet:"),
+        WidgetString (CFG_SECTION, "remote-source"),
+        {{remote_sources}}),
+    WidgetCheck (N_("Store fetched lyrics in local cache"),
+        WidgetBool (CFG_SECTION, "enable-cache")),
+    WidgetLabel (N_("<b>Local Storage</b>")),
+    WidgetCheck (N_("Load lyric files (.lrc) from local storage"),
+        WidgetBool (CFG_SECTION, "enable-file-provider"))
+};
```

Comment:
Could `defaults` and `widgets` be also shared between Qt and GTK somehow?

Reply:
Sure, you could put them in `preferences.h`. They can be `static` and don't need to be members of `Lyrics` or `LyricsQt`.